### PR TITLE
8286597: Implement PollerProvider on AIX

### PR DESCRIPTION
**Pull request by backwaterred**
This PR finalizes VThreads on AIX. After the heavy lifting was done by the lovely reinrich, all that remained was to implement the abstract methods in the Poller class. This was quite straightforward after I figured out that the Pollset library does not pick up on changes to the set of fds while blocked on a poll call.

### Notes

As mentioned above, the Pollset library won't recognize changes made to the set of fds while blocked on a call to `poll`. In order to allow the other threads to make changes to the pollset, I had to set a short (max 100ms) timeout for the call to poll in order to ensure that the pollset is refreshed. In my testing, this was a far more reliable solution than using a wakeup fd. 

I provided an empty implementation of two procedures in `src/hotspot/cpu/ppc/continuationHelper_ppc.inline.hpp` with some deductive work. Examining other implementations, I see they delegate this to `update_map_with_saved_link`. This is empty in `frame_ppc.inline.hpp`, so I concluded that there is nothing to do for the two continuationHelper procedures either. I wouldn't mind a second opinion on this choice.

The test `testSocketReadPeerClose2` in BlockingSocketOps was modified because setting so_linger doesn't cause an exception to be thrown on AIX. The underlying read call returns -1 as will other platforms, but instead of setting ECONNRESET, AIX just sets EAGAIN. I feel that this test may just not be feasible on AIX, and I've modified it accordingly.

I modified the timeout factor for the virtual/stress/Skynet.java test. This test passes without issue on my build & test system when run on its own, but when run as part of a test suite (eg. `make test TEST=jdk_loom`), it does not have enough time with the timeout previously specified in the test file.

### Testing

The following tests were performed on AIX.

- [x] T1 tests
- [x] hotspot_loom w/ -XX:+VerifyContinuations
- [x] jdk_loom w/ -XX:+VerifyContinuations

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8286597](https://bugs.openjdk.org/browse/JDK-8286597): Implement PollerProvider on AIX

### Reviewers
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**) ⚠️ Review applies to [8cf2249a](https://git.openjdk.org/jdk/pull/13452/files/8cf2249a71399cc547fda377b3e67097dbe5f3da)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)
 * [Vyom Tewari](https://openjdk.org/census#vtewari) (@vyommani - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13452/head:pull/13452` \
`$ git checkout pull/13452`

Update a local copy of the PR: \
`$ git checkout pull/13452` \
`$ git pull https://git.openjdk.org/jdk.git pull/13452/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13452`

View PR using the GUI difftool: \
`$ git pr show -t 13452`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13452.diff">https://git.openjdk.org/jdk/pull/13452.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13452#issuecomment-1538518829)

### CODE REVIEW

File: src/java.base/aix/classes/sun/nio/ch/PollsetProvider.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
```
Comment (offamitkumar) [inline src/java.base/aix/classes/sun/nio/ch/PollsetProvider.java:2]: Adding IBM header as well will be good, what do you think ? :-)

File: src/hotspot/cpu/ppc/continuationHelper_ppc.inline.hpp [hunk 2]
```diff
@@ -43,11 +43,11 @@ inline intptr_t* ContinuationHelper::frame_align_pointer(intptr_t* p) {
 
 template<typename FKind>
 inline void ContinuationHelper::update_register_map(const frame& f, RegisterMap* map) {
-  Unimplemented();
+  // Nothing to do
 }
 
 inline void ContinuationHelper::update_register_map_with_callee(const frame& f, RegisterMap* map) {
-  Unimplemented();
+  // Nothing to do
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/continuationHelper_ppc.inline.hpp:50]: Would it be better to call the empty `frame::update_map_with_saved_link` to be consistent with the other platforms? @reinrich: You may have an opinion.

File: src/java.base/share/classes/sun/nio/ch/Poller.java [hunk 3]
```diff
@@ -128,7 +128,7 @@ private void poll1(int fdVal, long nanos, BooleanSupplier supplier) throws IOExc
      * queues the file descriptor to the update thread, then parks until the file
      * descriptor is polled.
      */
-    private void poll2(int fdVal, long nanos, BooleanSupplier supplier) {
+    private void pollAsync(int fdVal, long nanos, BooleanSupplier supplier) {
```
Comment (AlanBateman) [inline src/java.base/share/classes/sun/nio/ch/Poller.java:131]: I don't object to renaming these private methods but "pollAsync" is confusing as the method is not asynchronously, both poll1 and poll2 are synchronous. This seems to be a drive-by change, maybe drop it from this PR as it's nothing to do with the port to AIX.

File: test/jdk/java/net/vthread/BlockingSocketOps.java [hunk 4]
```diff
@@ -191,7 +192,11 @@ void testSocketReadPeerClose2() throws Exception {
                 // read from s1 should block, then throw
                 try {
                     int n = s1.getInputStream().read();
-                    fail("read " + n);
+                    if (!Platform.isAix()) {
+                        fail("read " + n);
+                    } else {
+                        assertTrue(n == -1);
```
Comment (AlanBateman) [inline test/jdk/java/net/vthread/BlockingSocketOps.java:198]: This doesn't look right, read should not return -1 here.

File: src/java.base/share/classes/sun/nio/ch/Poller.java (hunk 3 above)
Comment (backwaterred) [reply src/java.base/share/classes/sun/nio/ch/Poller.java:131]: It's true that the change is not related to the implementation. But, I felt it was of benefit to change the poll1 and poll2 method names to be more descriptive. Since the changes already refactor the Poller implementation, I felt it was a good time to do it.

I'd be happy to change pollAsync to something that was more clear. How do you feel about `pollIndirect`?

Comment (AlanBateman) [reply src/java.base/share/classes/sun/nio/ch/Poller.java:131]: Changing to pollIndirect is okay if you really want to change these methods.

File: test/jdk/java/net/vthread/BlockingSocketOps.java (hunk 4 above)
Comment (backwaterred) [reply test/jdk/java/net/vthread/BlockingSocketOps.java:198]: I believe we get [here](https://github.com/openjdk/jdk/blob/master/src/java.base/unix/native/libnio/ch/SocketDispatcher.c#L44) with EAGAIN, but not ECONNRESET. So the -1 indicates that the read has failed.

My feeling is that the defined behaviour is not totally clear. [From setSockOpt](https://linux.die.net/man/3/setsockopt) (emphasis added by me):

> SO_LINGER
>   Lingers on a close() _if data is present_.

This does not define what happens if no data is present. In my testing, AIX behaved exactly like it does in `testSocketReadPeerClose1` so my understanding is that SO_LINGER had essentially no effect because there is no data waiting to be sent. As I see it, this test reduces to `testSocketReadPeerClose1` on AIX, so the test should be the same. Another option would be to skip it entirely on AIX.

Comment (AlanBateman) [reply test/jdk/java/net/vthread/BlockingSocketOps.java:198]: There are several tests in both the java/net and java/nio/channels tree that setup the conditions for a "hard reset", e.g. SocketChannel/ConnectionReset.java. I'm curious if these tests also fail on AIX.

Comment (backwaterred) [reply test/jdk/java/net/vthread/BlockingSocketOps.java:198]: Thanks for mentioning it. I'll take a look at those to see if they also fail.

Comment (backwaterred) [reply test/jdk/java/net/vthread/BlockingSocketOps.java:198]: That test passes. I'll take a look into the differences between the two tests.

Comment (AlanBateman) [reply test/jdk/java/net/vthread/BlockingSocketOps.java:198]: The long standing spec for SO_LINGER is "Enabling the option with a timeout of zero does a forceful close immediately". The wording isn't quite right but it is trying to say that if the timeout is set to zero then calling the close method will cause a forceful close. There are several tests that use this so there might be other failures on AIX.

File: src/hotspot/share/adlc/main.cpp [hunk 5]
```diff
@@ -229,6 +229,7 @@ int main(int argc, char *argv[])
   AD.addInclude(AD._CPP_file, "opto/regalloc.hpp");
   AD.addInclude(AD._CPP_file, "opto/regmask.hpp");
   AD.addInclude(AD._CPP_file, "opto/runtime.hpp");
+  AD.addInclude(AD._CPP_file, "runtime/continuation.hpp");
```
Comment (dholmes-ora) [inline src/hotspot/share/adlc/main.cpp:232]: This seems unrelated to the AIX changes. Is this include needed in general?

File: src/hotspot/cpu/ppc/continuationHelper_ppc.inline.hpp (hunk 2 above)
Comment (reinrich) [reply src/hotspot/cpu/ppc/continuationHelper_ppc.inline.hpp:50]: I'd prefer this version. It is clearer about the fact that `map` doesn't need to be updated on ppc with data from the callee `f`.

Comment (TheRealMDoerr) [reply src/hotspot/cpu/ppc/continuationHelper_ppc.inline.hpp:50]: Ok, I'm fine with it, but would prefer to have a comment describing this.

Comment (reinrich) [reply src/hotspot/cpu/ppc/continuationHelper_ppc.inline.hpp:50]: Or even: `Currently all registers are considered to be volatile and saved in the caller (java) frame if needed`.

File: src/hotspot/share/adlc/main.cpp (hunk 5 above)
Comment (backwaterred) [reply src/hotspot/share/adlc/main.cpp:232]: This change solved a build issue I had earlier in the development process. The build issue seems to have gone in my recent testing with this change removed. Thanks for pointing this out. I'll change the PR shortly.

File: test/jdk/java/net/vthread/BlockingSocketOps.java (hunk 4 above)
Comment (backwaterred) [reply test/jdk/java/net/vthread/BlockingSocketOps.java]: The ConnectionReset test you mentioned was very helpful. Comparing the pass there with the failure in `testSocketReadPeerClose2` led me to a better fix. I believe the test failure actually had to do with the order that the close, and setSoLinger methods were being executed in.

File: test/jdk/java/lang/Thread/virtual/stress/Skynet.java [hunk 6]
```diff
@@ -25,7 +25,7 @@
  * @test
  * @summary Stress test virtual threads with a variation of the Skynet 1M benchmark
  * @requires vm.continuations
- * @run main/othervm/timeout=300 -Xmx1g Skynet
+ * @run main/othervm/timeout=350 -Xmx1g Skynet
```
Comment (AlanBateman) [inline test/jdk/java/lang/Thread/virtual/stress/Skynet.java:28]: I assume this should be dropped from this PR as it is nothing to do with implement the AIX version of PollerProvider.

File: test/jdk/java/net/vthread/BlockingSocketOps.java [hunk 7]
```diff
@@ -59,6 +59,7 @@
 import java.net.SocketTimeoutException;
 
 import jdk.test.lib.thread.VThreadRunner;
+import jdk.test.lib.Platform;
```
Comment (AlanBateman) [inline test/jdk/java/net/vthread/BlockingSocketOps.java:62]: This seems to be left over from one of your previous iterations.

File: test/jdk/java/net/vthread/BlockingSocketOps.java [hunk 8]
```diff
@@ -185,8 +186,10 @@ void testSocketReadPeerClose2() throws Exception {
                 Socket s2 = connection.socket2();
 
                 // delayed abrupt close of s2
-                s2.setSoLinger(true, 0);
-                runAfterParkedAsync(s2::close);
+                runAfterParkedAsync(() -> {
+                    s2.setSoLinger(true, 0);
+                    s2.close();
+                });
```
Comment (AlanBateman) [inline test/jdk/java/net/vthread/BlockingSocketOps.java:192]: This is okay but shouldn't be necessary as the linger option was set before starting the thread to close the other end. The linger setting is per socket rather than per-thread so curious why the original was problematic on AIX.

File: test/jdk/java/lang/Thread/virtual/stress/Skynet.java (hunk 6 above)
Comment (backwaterred) [reply test/jdk/java/lang/Thread/virtual/stress/Skynet.java]: Agreed. It feels funny to push code I know will cause a test failure. But, I agree that it simplifies the git history, and aligns with the stated purpose of this PR. Thanks for the suggestion. I have made this change.

File: test/jdk/java/net/vthread/BlockingSocketOps.java (hunk 7 above)
Comment (backwaterred) [reply test/jdk/java/net/vthread/BlockingSocketOps.java]: Thanks for catching that. I made this change.

File: test/jdk/java/net/vthread/BlockingSocketOps.java (hunk 8 above)
Comment (backwaterred) [reply test/jdk/java/net/vthread/BlockingSocketOps.java]: This test was my last barrier to integration before I cleaned up the code. Now it seems that it is passing unmodified. So, I've reverted the change. I may see this failure again, but as you mentioned the test shouldn't need to be changed.

**Comment by backwaterred** [comment]
All comments have been addressed. This is a friendly request to please finalized your reviews.

**Review by TheRealMDoerr — APPROVED** [review]
Thanks for cleaning things up! LGTM. Maybe you would like to add a Copyright header for new files or basically rewritten files.

**Comment by AlanBateman** [comment]
I'm not involved in the AIX port, and have not used pollset, but I am puzzled by PollsetProvider as I expected it to be named Pollset (its not a factory/provider of Pollset, it instead provides an interface to the pollset I/O facility). If you look at the naming/architecture for the other platforms then you'll see what I mean.

File: src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java [hunk 9]
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2022, 2023, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2023, IBM Corp.
```
Comment (offamitkumar) [inline src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java:3]: Maybe this is correct, but still you may want to take a look at it: https://www.ibm.com/docs/en/security-verify?topic=information-copyright-statement

Shouldn't the header look like this `© Copyright IBM Corporation 2023, `

Comment (backwaterred) [reply src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java:3]: Hi @offamitkumar, thanks for the review. Back when I did the JFR changes, I contacted IBM legal to get info on the copyright header that I should use. This was their suggestion.

**Comment by TheRealMDoerr** [comment]
> I'm not involved in the AIX port, and have not used pollset, but I am puzzled by PollsetProvider as I expected it to be named Pollset (its not a factory/provider of Pollset, it instead provides an interface to the pollset I/O facility). If you look at the naming/architecture for the other platforms then you'll see what I mean.

That's a valid point. `Pollset` may be a better name.

**Comment by backwaterred** [comment]
@AlanBateman, I believe you are active in the various Poller implementations, and have looked at these changes pretty thoroughly. Would you be comfortable finalizing your review?

File: src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java [hunk 10]
```diff
@@ -25,39 +26,86 @@
...
+        return n;
+    }
+
+    int pollInner(int subInterval) throws IOException {
+        // The poll loop may start polling before any fds have been registered. But, if we use set
+        // size 0 to allocatePollArray, it will return the null address and pollsetPoll will complain.
+        // We avoid that by just passing set size 1, and letting poll block for subInterval.
+        long buffer = Pollset.allocatePollArray(setsize > 0 ? setsize : 1);
+        int n = Pollset.pollsetPoll(setid, buffer, setsize, subInterval);
+        for(int i=0; i<n; i++) {
```
Comment (turbanoff) [inline src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java:102]:
```suggestion
        for (int i=0; i<n; i++) {
```

**Comment by backwaterred** [comment]
Thanks for the suggestions @turbanoff. I've made those changes.

File: src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java [hunk 11]
```diff
@@ -25,39 +26,86 @@
...
+        switch (timeout) {
+            case 0:
+                n = pollInner(0);
+                break;
+            case Pollset.PS_NO_TIMEOUT:
+                do { n = pollInner(100); } while (n == 0);
+                break;
+            default:
+                Instant end = Instant.now().plusMillis(timeout);
+                do { n = pollInner(100); } while (n == 0 && Instant.now().isBefore(end));
```
Comment (AlanBateman) [inline src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java:90]: Now, the Poller uses poll(-1) to poll indefinitely so the 0/default cases aren't used. If we do start to use the timed case then L90 probably should probably be optimized to avoid Instant.now. Also, probably should be reformatted to make the do-while loop a bit easier to read.

File: src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java [hunk 12]
```diff
@@ -25,39 +26,86 @@
...
+        // size 0 to allocatePollArray, it will return the null address and pollsetPoll will complain.
+        // We avoid that by just passing set size 1, and letting poll block for subInterval.
+        long buffer = Pollset.allocatePollArray(setsize > 0 ? setsize : 1);
+        int n = Pollset.pollsetPoll(setid, buffer, setsize, subInterval);
+        for (int i=0; i<n; i++) {
+            long eventAddress = Pollset.getEvent(buffer, i);
+            int fd = Pollset.getDescriptor(eventAddress);
+            polled(fd);
+        }
+        Pollset.freePollArray(buffer);
```
Comment (AlanBateman) [inline src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java:107]: If I read this correctly, there is a malloc/free per poll, is that right?

Comment (AlanBateman) [reply src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java:107]: If you look at the Poller implementations on the other platform you'll see that they allocate the poll array at construction time, I think that is what you want here too.

Comment (backwaterred) [reply src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java:107]: You are correct that there is a malloc/free per poll. I made this decision intentionally.

The other implementations allocate a fixed size array at object creation. This imposes an upper limit on the number of fds that can be polled. I made the decision to move this allocation to the poll call and track the size of the poll set. This way, (1) the only limit to the size of the poll set is the system limit on number of open fds (if you can open an fd it can be polled) (2) there is no wasted space to an arbitrarily large array when there are only a few fds polled.

File: src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java (hunk 11 above)
Comment (backwaterred) [reply src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java:90]: I'm happy to tweak this case. In terms of formatting, would you prefer?

```
do { n = pollInner(100); }
while (n == 0 && Instant.now().isBefore(end));
```

or the more conventional

```
do {
  n = pollInner(100);
} while (n == 0 && Instant.now().isBefore(end));
```

or another idea?

---

In regards to `Instant.now(...)`. I'm not sure I understand your suggestion. Could you please clarify? Is there a different time-object I should use instead?

File: src/java.base/aix/classes/sun/nio/ch/Pollset.java [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+     */
+    public static void freePollArray(long address) {
+        unsafe.freeMemory(address);
+    }
+
+    /**
+     * Returns event[i];
+     */
+    public static long getEvent(long address, int i) {
+        return address + (SIZEOF_POLLFD*i);
```
[the remark this answers is not in the document]
Comment (backwaterred) [reply src/java.base/aix/classes/sun/nio/ch/Pollset.java]: Thanks. Fixed :smile:.

File: src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java (hunk 12 above)
Comment (vyommani) [reply src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java:107]: there will be some performance impact  in "allocation/de-allocation"  memory  per poll approach.

Comment (AlanBateman) [reply src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java:107]: > The other implementations allocate a fixed size array at object creation. This imposes an upper limit on the number of fds that can be polled. I made the decision to move this allocation to the poll call and track the size of the poll set. This way, (1) the only limit to the size of the poll set is the system limit on number of open fds (if you can open an fd it can be polled) (2) there is no wasted space to an arbitrarily large array when there are only a few fds polled.

I don't have any experience on AIX but docs page for pollset_poll suggests it populates the array with the file descriptors that have pending events. If I read this correctly, then it's nothing to do with the number of file descriptors registered and it works similarly to /dev/poll, epoll and kqueue. So I think maybe look at this again.

Comment (backwaterred) [reply src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java:107]: In regards to the speed and fragmentation issues. I think these are compelling reasons to consider the other approach (Option 2 below).

With respect to memory being returned to the OS: I would expect this issue to be minimal because the memory would 'stick' to the process and just be recycled.

> If I read this correctly, then it's nothing to do with the number of file descriptors registered

The implementation as it is written keeps track of the number of fds registered and only allocates memory for that number (0-2 in my testing) when the poll starts. The other implementations allocate space for 512 pollfds at object creation. You are correct that the library only _populates_ the space if there is an event, but space has to be allocated before hand.

The memory requirements per poller are actually doubled because there is always a reader and writer created.

### Option 1: Memory allocated per-poll

Pros:
- Memory footprint is minimal ~ `1 * sizeof(struct pollfd)`.

Cons:
- Possible performance issues (speed).
- Possible memory fragmentation issues.

### Option 2: Memory allocated on Poller creation

Pros:
- No issues with memory fragmentation or allocation performance (since malloc/free calls are minimized).

Cons:
- Arbitrary upper limit on number of events per poll.
- Much larger memory footprint `512 * sizeof(struct pollfd)`

---

Option 2 seems to be the popular choice, and I'm amenable to changing that. It does seems a bit silly to allocate 2 * 512 * 8 bytes == 8 KiB for each Poller when I've only seen 4 * 8 bytes in use.

Comment (TheRealMDoerr) [reply src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java:107]: Are there alternatives like using alloca?

Comment (backwaterred) [reply src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java:107]: > Are there alternatives like using alloca?

Cool suggestion! I actually hadn't heard of it before. I am still not totally sure how it's different from just using local variables which should be on the stack anyway. From a quick man-page search it looks viable to do this here.

Another possibility: I could use Option 2, but not use such a large value for MAX_FDS. The max we choose just dictates how many results are returned by each call to pollset_poll. Any fds not returned in one call should just be returned the next call, and the poll loop is constantly making poll calls.

Some quick testing suggests that pollset does a pretty good job of returning unseen events first. Polling a set of 8 files that could all be read with a pollset size of 1, I saw 8 unique fds.

So, I'm leaning towards Option 2 with MAX_FDS of ~16. Does that seems like a reasonable compromise?

Comment (AlanBateman) [reply src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java:107]: > Option 2 seems to be the popular choice, and I'm amenable to changing that. It does seems a bit silly to allocate 2 * 512 * 8 bytes == 8 KiB for each Poller when I've only seen 4 * 8 bytes in use.

There are 2 Poller threads per VM. They run in a loop that does a blocking poll (pollset_poll in the discussion here) to pick up events in bulk. 8K doesn't seem too bad and you could reduce the 512 to a small batch size if you want.

Another point is that there is configuration knob to control the registration. In "direct" mode, virtual threads that need to block on I/O will will attempt to directly arm the file descriptor. This means that implRegister will be called concurrently by many threads, the implRegister implementation in the changes here will increment and decrement setsize without synchronization. You may not have seen this as the default uses "indirect" mode.

File: src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java [hunk 14]
```diff
@@ -25,39 +26,79 @@
...
  * Poller implementation based on the AIX Pollset library.
  */
 
 class PollsetPoller extends Poller {
 
+    static { Pollset.init(); /* Dynamically loads pollset C functions */ }
+
+    private final int setid;
+    private final long pollBuffer;
+    private static final int POLL_BUFFER_SIZE = 16;
```
Comment (AlanBateman) [inline src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java:42]: 16 seems very low. On the other platforms the equivalent is sized to read up to 256 or 512 events in one syscall but maybe there is a reason to choose this small number? You might need to run Helidon Nima or some other server using virtual threads to see the impact on this.

The name used on other platforms for the constant is "MAX_EVENTS_TO_POLL" to better reflect the usage.

File: src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java (hunk 12 above)
Comment (TheRealMDoerr) [reply src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java]: I agree. 8K per Poller threads sounds still affordable.

File: src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java (hunk 14 above)
Comment (vyommani) [reply src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java:42]: I agree with Alan, when I was in IBM, I re-implemented the same for IBM Z/OS and my initial buffer size was very low. The product team complains that first 512 clients connections  were slow. I increased buffer size to 512( MAX_POLL_EVENTS = 512;) and problem got resolved. Although  Z/OS implementation was based on “select and poll” API which is different than your AIX implementation and my be, buffer size is not applicable to AIX.

File: src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java (hunk 12 above)
Comment (backwaterred) [reply src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java]: Ok. I have changed the implementation to be similar to that of the other platforms--including the size of the buffer.

File: src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java (hunk 14 above)
Comment (backwaterred) [reply src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java]: Thanks for your feedback. I have changed this to reference the existing MAX_POLL_EVENTS constant which is set to 512.

File: src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java (hunk 11 above)
Comment (backwaterred) [reply src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java:87]: Just waiting for clarification from you on this discussion @AlanBateman.

File: src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java [hunk 15]
```diff
@@ -25,39 +26,78 @@
...
+        switch (timeout) {
+            case 0:
+                n = pollInner(0);
+                break;
+            case Pollset.PS_NO_TIMEOUT:
+                do { n = pollInner(100); } while (n == 0);
+                break;
+            default:
+                Instant end = Instant.now().plusMillis(timeout);
+                do { n = pollInner(100); } while (n == 0 && Instant.now().isBefore(end));
```
Comment (AlanBateman) [inline src/java.base/aix/classes/sun/nio/ch/PollsetPoller.java:87]: I don't have a strong opinion on that as it's a code path that isn't used by the Poller implementation at this time.

File: src/java.base/aix/classes/sun/nio/ch/Pollset.java [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    public static final int PS_POLLPRI = 0x4;
+
+    // revent errcodes
+    public static final char PS_POLLNVAL = 0x8000;
+    public static final char PS_POLLERR  = 0x4000;
+
+    public static final int PS_NO_TIMEOUT = -1;
+
+    // maximum number of events to poll at a time
+    public static final int MAX_POLL_EVENTS = 512;
```
[the remark this answers is not in the document]
Comment (backwaterred) [reply src/java.base/aix/classes/sun/nio/ch/Pollset.java]: Fair comment. Thanks for pointing that out. I have made this change.

**Review by AlanBateman — APPROVED** [review]
I don't have any other comments on this change.

**Review by vyommani — APPROVED** [review]
looks ok to me.

**Comment by backwaterred** [comment]
Thanks for the reviews, comments, and feedback.

/integrate
